### ISSUE TITLE
com.google.android.exoplayer2.source.UnrecognizedInputFormatException: None of the available extractors

### ISSUE THREAD

**Issue by jhonatanluiz** [ryanheise/just_audio#315]
**Which API doesn't behave as documented, and how does it misbehave?**
Error when playing simple mp3 file in assets.

**Minimal reproduction project**

    AudioPlayer player = new AudioPlayer();
    player.setAsset('sound/singlebeep.mp3');
    player.play();

**To Reproduce (i.e. user steps, not code)**
(n\a)

**Error messages**

```
   com.google.android.exoplayer2.ExoPlaybackException: Source error
       at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:554)
       at android.os.Handler.dispatchMessage(Handler.java:103)
       at android.os.Looper.loop(Looper.java:214)
       at android.os.HandlerThread.run(HandlerThread.java:67)
   Caused by: com.google.android.exoplayer2.source.UnrecognizedInputFormatException: None of the available extractors (Mp3Extractor, FlvExtractor, FlacExtractor, WavExtractor, FragmentedMp4Extractor,       at com.google.android.exoplayer2.source.BundledExtractorsAdapter.init(BundledExtractorsAdapter.java:92)
       at com.google.android.exoplayer2.source.ProgressiveMediaPeriod$ExtractingLoadable.load(ProgressiveMediaPeriod.java:1028)
       at com.google.android.exoplayer2.upstream.Loader$LoadTask.run(Loader.java:415)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
       at java.lang.Thread.run(Thread.java:919)
E/AudioPlayer(13513): TYPE_SOURCE: None of the available extractors (Mp3Extractor, FlvExtractor, FlacExtractor, WavExtractor, FragmentedMp4Extractor, Mp4Extractor, AmrExtractor, PsExtractor, OggExtractor, TsExtractor, MatroskaExtractor, AdtsExtractor, Ac3Extractor, Ac4Extractor) could read the stream.
```

**Expected behavior**
Play sound mp3 file in assets.

**Screenshots**
(n/a)

**Desktop (please complete the following information):**
 - OS: MacOS BigSur 11.2.1
 - Browser [e.g. chrome, safari + version]

**Smartphone (please complete the following information):**
 - Device: Galaxy J7
 - OS: Android 10

**Flutter SDK version**
```
[✓] Flutter (Channel stable, 1.22.6, on macOS 11.2.1 20D74 darwin-x64, locale pt-BR)
    • Flutter version 1.22.6 at /Users/jhonatanluiz/flutter
    • Framework revision 9b2d32b605 (4 weeks ago), 2021-01-22 14:36:39 -0800
    • Engine revision 2f0af37152
    • Dart version 2.10.5

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/jhonatanluiz/Library/Android/sdk
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = /Users/jhonatanluiz/Library/Android/sdk
    • Java binary at: /Users/jhonatanluiz/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/201.7042882/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.9.3

[!] Android Studio (version 4.1)
    • Android Studio at /Users/jhonatanluiz/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/201.7042882/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] Connected device (1 available)
    • Android SDK built for x86 (mobile) • emulator-5554 • android-x86 • Android 10 (API 29) (emulator)```

**Additional context**
Add any other context about the problem here.

**Comment by ryanheise** [comment]
Hi @jhonatanluiz  I am not able to reproduce the problem with what you have given me. Can you provide me a reproduction case?

**Comment by tiloc** [comment]
This is looking similar/identical to issue #299 which I posted a while ago. While the message was mentioning Exoplayer and formats, it ultimately turned out to be a missing `await` in front of the `player.setAsset('sound/singlebeep.mp3');`. This seems to result in broken data being fed to player.play(). I changed my code like the following to make it work:

```
final player = new AudioPlayer();
await player.setAsset('sound/singlebeep.mp3');
unawaited(player.play());
```

**Comment by ryanheise** [comment]
@tiloc that's an interesting find. Internally, just_audio copies the asset to a file and then ExoPlayer attempts to read from that file. So based on that theory, ExoPlayer is attempting to play from the file before the file has actually finished being written.

Looking at the code, I don't actually see how this situation could happen with the above sequence. Even without await, ExoPlayer is not given the location of the file until after it has been fully written.

The one exception is if another instance of `AudioPlayer` in parallel tries to load the same asset before the first player has completely written the file. In this case, the second player will see that the file already exists (without checking that it's not actually complete) and then immediately pass that onto ExoPlayer. That's a bug which I can fix.

But I'm not sure how a single player with `setAsset` called once could cause the issue. A reproduction project would help to investigate it.

**Comment by MobileMon** [comment]
I'm having the same issue. Can't get this library to work at all

**Comment by ryanheise** [comment]
Are you loading from a local asset? Otherwise can you provide a link to the audio that fails?

**Comment by ryanheise** [comment]
Since the original issue above has no reproduction project, I'll close this. @tiloc I would be interested to fix that `setAsset` issue if you are able to open a new issue with a minimal reproduction project.
